btx/processing/rawimagetimetool.py

```diff
@@ -0,0 +1,575 @@
+    avail = np.array([int(a) for a in avail])
+    sort_idx = np.argsort(avail)
+    idx = np.searchsorted(avail[sort_idx], run, side='right') - 1
+    return fnames[sort_idx[idx]]
+    try:
+        return fnames[sort_idx[idx]]
+    except IndexError:
+        print('File not found.')
+        return ''
+
+class RawImageTimeTool:
+    """! Class to reimplement time tool analysis at LCLS from raw images.
+
+    Uses psana to interface with experimental data to retrieve raw camera images.
+    Edge detection is implemented in order to perform jitter correction.
+
+    Properties:
+    -----------
+    model - Retrieve or load polynomial model coefficients.
+
+    Methods:
+    --------
+    __init__(self, expmt: str, savedir: str) - Instantiate an analysis object for experiment expmt.
+    open_run(self, run: str) - Open DataSource for a specified run.
+    get_tt_name(self) - Determine time tool camera name in data files.
+    calibrate(self, run: str, order: int, figs=True) - Calibrate the analysis object for jitter correction on specified run.
+    process_run(self) - Perform edge detection on all events in a run.
+    detect_edge(self, img: np.array, kernel: np.array) - Perform edge detection on an image.
+    crop_image(self, img: np.array) - Crop time tool image to ROI.
+    fit_calib(self, delays: np.array, edges: np.array, amplitudes: np.array,
+            fwhm: np.array = None, order: int = 2) - Fit polynomial to calibration run data.
+    ttstage_code(self, hutch) - Return the time tool target time stage for a given hutch.(
+    edge_to_time(self, edge)
+    actual_time(self, edge, nominal)
+    timetool_correct(self, edge, nominal, model, figs=True)
+    plot_calib
+    plot_hist
+    """
+    # Class vars
+    ############################################################################
+
+    # Methods
+    ############################################################################
+
+    def __init__(self, expmt: str, savedir: str):
+        """! Initializer for the time tool analysis object.
+
+        @param expmt (str) Experiment accession name. E.g. cxilr1234.
+        @param savedir (str) Directory where any output files are saved.
+        """
+        ## @var expmt
+        # (str) Experiment accession name
+        self.expmt = expmt
+
+        ## @var hutch
+        # (str) Experimental hutch. Extracted from expmt. Needed for camera/stage accession.
+        self.hutch = self.expmt[:3]
+
+        ## @var _model
+        # (list) List of calibration model polynomial coefficients. Empty if not calibrated.
+        self._model = []
+
+        ## @var savedir
+        # (str) Directory to save output files to. Figures may be placed in a subdirectory of this one.
+        self.savedir = savedir
+
+    def open_run(self, run: str):
+        """! Open the psana DataSource and time tool detector for the specified
+        run. This method MUST be called before calibration, analysis etc. The
```

Comment:
Maybe call it in `__init__()` to ensure it's always called before calibration, analysis, etc?

Reply:
Ah I see why this is decoupled from the `__init__` given that there are two different kinds of runs: calibration and analysis.

How about we do the following: `__init__()` takes two additional inputs: `self._run_calib` and `self._run_analysis` or something like that?